### ISSUE TITLE
Events should be handled asynchronously to avoid deadlocks

### ISSUE THREAD

**Issue by codablock** [SUSE/Portus#1235]
I ran into the same problem described here: https://github.com/SUSE/Portus/issues/940

As I understand it (based on my own debugging), the problem is that when events come in, the processing of these results in requests to the registry to get some additional information about the images being pushed. When this happens, the registry returns 401 to indicate a missing token. Portus then acts with an authorization attempt by doing a request to itself. This request however does not finish as for some reason Portus has to wait for the current request to finish, which is never happening as it's waiting for the auth request to finish.

I do not understand why Portus can not handle multiple requests at the same time here. I have setup puma to use multiple workers and threads, so it should be able to handle multiple requests. Maybe there is something in the code that causes database locking, but I'm not familiar enough with Ruby to debug this properly.

Nevertheless, my feeling is that the events processing should actually happen asynchronously to avoid any type of possible deadlock or request performance degradation. What should happen is:
1. Registry notifies Portus about events
2. Events endpoint of Portus gets called, entering the request handler
3. Add the event to something that is capable of background/asynchronously processing
4. Return from the event request handler immediately
5. Let the background processing do its work

An alternative idea to avoid the deadlock at least for now is to somehow call the token generation code directly instead of going through a HTTP request. If the deadlock is caused by a limitation on parallel requests, this would fix it. Also, if it's happening due to DB locks, this would probably also fix it as it could happen in the same transaction. This however feels like a dirty trick as it kind of ignores the registries auth settings.

**Comment by codablock** [comment]
After some more investigation, I figured out that the manifest fetching was actually a bug. It tried to fetch the manifest even though the required info (digest) was already in the event. A fix to this is https://github.com/SUSE/Portus/issues/1236

Also, after reading through details how Ruby on Rails handles concurrency, I figured out that there are some connections between these issues and concurrency handling. As I understand it, RoR disables concurrency in development environments with the help of a mutes around all request handling, so that only one request can be served at the same time. I'm not sure how production is different in this behavior.  

Also, at least for development, caching of classes is disabled, resulting in the cached token property of the Registry model to being reused, increasing the number of needed authentication calls. So the behavior with these issues seen in development may be different from what you see in production.

Disclaimer: I'm not a Ruby/RoR coder, so my understanding of RoR internals may be wrong. I base my assumptions on the experience from other frameworks in other languages.
